File: reference.net/db4o-reference-chapters/src/Db4objects.Db4odoc/reflections/LoggingArray.cs
/* Copyright (C) 2005   db4objects Inc.   http://www.db4o.com */

using System;
using Db4objects.Db4o.Reflect;
using Db4objects.Db4o.Reflect.Net;

namespace Db4objects.Db4odoc.Reflections
{
	public class LoggingArray : Db4objects.Db4o.Reflect.IReflectArray
	{
		private readonly Db4objects.Db4o.Reflect.IReflector _reflector;

		internal LoggingArray(Db4objects.Db4o.Reflect.IReflector reflector)
		{
			_reflector = reflector;
		}

		public virtual int[] Dimensions(object arr)
		{
			return new int[] { GetLength(arr) };
		}

		public virtual int Flatten(object a_shaped, int[] a_dimensions, int a_currentDimension
			, object[] a_flat, int a_flatElement)
		{
			object[] shaped = (object[])a_shaped;
			System.Array.Copy(shaped, 0, a_flat, 0, shaped.Length);
			return shaped.Length;
		}

		public virtual object Get(object onArray, int index)
		{
			return ((object[])onArray)[index];
		}

        public virtual Db4objects.Db4o.Reflect.IReflectClass GetComponentType(Db4objects.Db4o.Reflect.IReflectClass a_class)
		{
			while (a_class.IsArray()) 
			{
				a_class = a_class.GetComponentType();
			}
			return a_class;
		}

		public virtual int GetLength(object array)
		{
			return ((object[])array).Length;
		}

		public virtual bool IsNDimensional(Db4objects.Db4o.Reflect.IReflectClass a_class)
		{
			return false;
		}

		private static Type GetNetType(IReflectClass a_class)
		{
			return ((NetClass)a_class).GetNetType();
		}

		public virtual object NewInstance(Db4objects.Db4o.Reflect.IReflectClass componentType, int
			length)
		{
			return System.Array.CreateInstance(GetNetType(componentType), length);
		}

		public virtual object NewInstance(Db4objects.Db4o.Reflect.IReflectClass componentType, int[]
			dimensions)
		{
			return NewInstance(componentType, dimensions[0]);
		}

		public virtual void Set(object onArray, int index, object element)
		{
			((object[])onArray)[index] = element;
			return;
		}

		public virtual int Shape(object[] a_flat, int a_flatElement, object a_shaped, int[]
			a_dimensions, int a_currentDimension)
		{
			object[] shaped = (object[])a_shaped;
			System.Array.Copy(a_flat, 0, shaped, 0, a_flat.Length);
			return a_flat.Length;
		}
	}
}

File: reference.vb.net/Db4objects.Db4odoc/Reflections/LoggingArray.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 

Imports System
Imports Db4objects.Db4o.Reflect
Imports Db4objects.Db4o.Reflect.Net

Namespace Db4objects.Db4odoc.Reflections
    Public Class LoggingArray
        Implements IReflectArray
        Private ReadOnly _reflector As IReflector

        Public Sub New(ByVal pReflector As IReflector)
            _reflector = pReflector
        End Sub

        Public Function Dimensions(ByVal arr As Object) As Integer() Implements IReflectArray.Dimensions
            Return New Integer() {GetLength(arr)}
        End Function

        Function [Get](ByVal onArray As Object, ByVal index As Integer) As Object Implements IReflectArray.Get
            Return (CType(onArray, Object()))(index)
        End Function

        Function Flatten(ByVal a_shaped As Object, ByVal a_dimensions() As Integer, ByVal a_currentDimension As Integer, ByVal a_flat() As Object, ByVal a_flatElement As Integer) As Integer Implements IReflectArray.Flatten
            Dim shaped() As Object = CType(a_shaped, Object())
            System.Array.Copy(shaped, 0, a_flat, 0, shaped.Length)
            Return shaped.Length
        End Function

        Function GetComponentType(ByVal a_class As IReflectClass) As IReflectClass Implements IReflectArray.GetComponentType
            While a_class.IsArray()
                a_class = a_class.GetComponentType()
            End While
            Return a_class
        End Function

        Public Function GetLength(ByVal array As Object) As Integer Implements IReflectArray.GetLength
            Return (CType(array, Object())).Length
        End Function

        Function IsNDimensional(ByVal a_class As IReflectClass) As Boolean Implements IReflectArray.IsNDimensional
            Return False
        End Function

        Function GetNetType(ByVal a_class As IReflectClass) As Type
            Return (CType(a_class, NetClass)).GetNetType()
        End Function

        Function NewInstance(ByVal componentType As IReflectClass, ByVal dimensions As Integer()) As Object Implements IReflectArray.NewInstance
            Return NewInstance(componentType, dimensions(0))
        End Function


        Function NewInstance(ByVal componentType As IReflectClass, ByVal length As Integer) As Object Implements IReflectArray.NewInstance
            Return System.Array.CreateInstance(GetNetType(componentType), length)
        End Function


        Sub [Set](ByVal onArray As Object, ByVal index As Integer, ByVal element As Object) Implements IReflectArray.Set
            Dim shaped() As Object = CType(onArray, Object())
            shaped.SetValue(element, index)
            Return
        End Sub

        Function Shape(ByVal a_flat() As Object, ByVal a_flatElement As Integer, ByVal a_shaped As Object, ByVal a_dimensions() As Integer, ByVal a_currentDimension As Integer) As Integer Implements IReflectArray.Shape
            Dim shaped() As Object = CType(a_shaped, Object())
            System.Array.Copy(a_flat, 0, shaped, 0, a_flat.Length)
            Return a_flat.Length
        End Function
    End Class
End Namespace